src/jedi_nft.cairo

```diff
@@ -0,0 +1,490 @@
+            return self._merkle_roots.read(task_id);
+        }
+
+        fn mint_whitelist(
+            ref self: ContractState,
+            token_id: u128,
+            proof: Array<felt252>,
+            token_metadata: TokenMetadata
+        ) {
+            let caller = get_caller_address();
+            let merkle_root = self._merkle_roots.read(token_metadata.task_id);
+            assert(merkle_root != 0, 'merkle root not set');
+            let mut leaf: felt252 = LegacyHash::hash(caller.into(), token_id);
+            leaf = LegacyHash::hash(leaf, token_metadata.task_id);
+            leaf = LegacyHash::hash(leaf, token_metadata.name);
+            leaf = LegacyHash::hash(leaf, token_metadata.rank);
+            leaf = LegacyHash::hash(leaf, token_metadata.score);
+            leaf = LegacyHash::hash(leaf, token_metadata.level);
+            leaf = LegacyHash::hash(leaf, token_metadata.total_eligible_users);
+
+            let mut merkle_tree = MerkleTreeTrait::new();
+            let result = merkle_tree.verify(merkle_root, leaf, proof.span());
+            assert(result == true, 'verify failed');
+
+            self._token_metadata.write(token_id, token_metadata);
+            let is_minted = self._completed_tasks.read((token_metadata.task_id, caller));
+            assert(!is_minted, 'ALREADY_MINTED');
+            self._completed_tasks.write((token_metadata.task_id, caller), true);
+            let mut erc721_self = ERC721::unsafe_new_contract_state();
+            erc721_self._mint(to: caller, token_id: token_id.into());
+        }
+
+        fn get_token_metadata(self: @ContractState, token_id: u128) -> TokenMetadata {
+            return self._token_metadata.read(token_id);
+        }
+
+        fn set_mint_sig_pub_key(ref self: ContractState, mint_sig_public_key: felt252) {
+            self._only_owner();
+            self._mint_sig_public_key.write(mint_sig_public_key);
+        }
+
+        fn get_mint_sig_pub_key(self: @ContractState) -> felt252 {
+            return self._mint_sig_public_key.read();
+        }
+
+        fn mint_sig(
+            ref self: ContractState,
+            token_id: u128,
+            signature: Span<felt252>,
+            token_metadata: TokenMetadata
+        ) {
+            let mint_sig_public_key = self._mint_sig_public_key.read();
+            assert(mint_sig_public_key != 0, 'MINT_SIG_PUBLIC_KEY_NOT_SET');
+            let caller = get_caller_address();
+            let mut hashed = LegacyHash::hash(caller.into(), token_id);
+            hashed = LegacyHash::hash(hashed, token_metadata.task_id);
+            hashed = LegacyHash::hash(hashed, token_metadata.name);
+            hashed = LegacyHash::hash(hashed, token_metadata.rank);
+            hashed = LegacyHash::hash(hashed, token_metadata.score);
+            hashed = LegacyHash::hash(hashed, token_metadata.level);
+            hashed = LegacyHash::hash(hashed, token_metadata.total_eligible_users);
```

Comment:
Should we add chain_id to this hash? If not, does it pose any risk that testnet signature can be used on mainnet?

Reply:
no need, we already add user address as part of the hash, the caller must be user, not matter which chain